GVFS/GVFS.Common/ProductUpgrader.cs

```diff
@@ -0,0 +1,38 @@
+using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.IO;
+
+namespace GVFS.Common
+{
+    public partial class ProductUpgrader
+    {
+        public static IProductUpgrader CreateUpgrader(ITracer tracer, out string error)
+        {
+            IProductUpgrader upgrader;
+            bool isEnabled;
+            bool isConfigured;
+            error = string.Empty;
+            
+            if ((upgrader = GitHubUpgrader.Create(tracer, out isEnabled, out isConfigured)) != null)

```

Comment:
GitHubUpgrader constructor now takes as GitHubUpgraderConfig as an argument. This new config class has been created so that in future NugetUpgrader can have its own NugetUpgraderConfig. The static Create method, helps keep the Upgrader specific Configuration objects private and within the respective Upgrader classes. The Factory method (`ProductUpgrader.TryCreateUpgrader()`) does not need to know details of GitHub or Nuget Upgrader Configs.